### ISSUE TITLE
Particles - Selecting particles resets its scale

### ISSUE THREAD

**Issue by dllllllllllb** [TheGameCreators/GameGuruRepo#2341]
After changing particles scale, the scale gets reset when the particle is selected again.
https://gyazo.com/fc595fddfae4e97ceafd8b34ce8c3001

I tried looking into this yesterday but I couldn't fix it. I think this has something to do with deleting and creating a new entity when its "moved".

**Comment by dllllllllllb** [comment]
@LeeBamberTGC So this one seems more complicated than it looks. I don't think the problem is with the UI reading wrong values as other objects work perfectly fine. To make sure, I commented out the code that is responsible for scale editing and selecting a particle still reset its scale.

Then I tried looking into other causes of this and thought that resetting the widget scale after resizing the particle is the issue but I could not find all the code responsible for resetting widgets scale. When selecting another entity, the widget scale would reset, but particle size stayed the same.

Then I looked into the code responsible for selecting an entity and I though this is the issue. I commented out appropriate gridedit_addentitytomap() function, and after selecting the particle, the widget got deleted but the particles stayed as they were.

I then looked where the scale values are assigned in the entity_addentitytomap_core() function and changed them, but they were not responsible for this issue. 

I think that the widget is the cause of the problem. When you select a particle, its widget gets updated to have it's original scale which then gets passed on to the particles. I tried looking for this, but I couldn't find all the code responsible. However, I changed the t.entityprofile[entid].scale value and selecting a particle changes the widget scale once selected, and the scale of particles with it :]

TL;DR - t.entityprofile[entid].scale that is used to reset the widget scale also impacts the particle scale.

**Comment by LeeBamberTGC** [comment]
@dllllllllllb See if you can apply a fix based on what you have learned, and we will see if there are any knock-on effects, thanks!

**Comment by dllllllllllb** [comment]
@LeeBamberTGC I managed to pin point where the issue is, I tried a couple of things but I can't come up with a good solution. 
Here is the issue:
![image](https://user-images.githubusercontent.com/61796872/158578446-15958557-76f7-4581-8810-ebb107dea1e9.png)
The size of the particles is dependent on the size of the widget, but when you select the widget it successfully resets to it's base size and keeps particles size the same due to the scaling code in M-GridEdit. I tried to apply that logic to where the issue is but with no success. 

If I don't reset the widget scale, particle scale stays the same, but the widget scales with the particles if it's not selected.
https://gyazo.com/1a679032eef8f06d079aa096e7adad5d

I'm not sure what to do here.

@ZakJudges @PaulSJ Do you have any ideas I should try? Maybe there is a way to separately scale the widget?

**Comment by ZakJudges** [comment]
@dllllllllllb 

I haven't looked at where entity_positionandscale() is being called, but I suspect the issue is with this line:

`float fScale = t.entityprofile[t.entityelement[t.tte].bankindex].scale;`

Keep in mind that t.entityprofile[] contains the master object data (what is in the .fpe file).

t.entityelement[] contains data specific to each instance of an object in the world.

So when you say float fScale = t.entityprofile[...].scale, what you are saying is scale should be set to the original master object's scale.

**Comment by dllllllllllb** [comment]
@ZakJudges That's the issue, when we change the scale of the particles, we only want to change the scale the particles and not the widget, but particles share the scale with the widget. 

When you select an object, entity_positionandscale() is called, initial ScaleObject() function call sets the particles scale to what they should be, and then it get's reset below to make sure the widget stays the original size.

In M-GridEdit, the same code is used to reset the widget, but particles somehow don't change their scale after the widget is reset.
![image](https://user-images.githubusercontent.com/61796872/158589294-45477739-43d7-4d82-9553-f978f6ae3559.png)

If you comment out the underlined function, and scale the particles, both particles and the widget change scale.

**Comment by ZakJudges** [comment]
@dllllllllllb Ah sorry, I read the original post wrong. When you move the particle (and it gets added to the map again), I would imagine that the code responsible for sending the particle data to the gpu is also called. My guess is that it grabs the scale of the object (which was made smaller for the widget), and then applies it to the particle effect.

You could check entity_prepareobj() and see if it is called when you move the particle. I believe there is code in there that updates the particle. It could be the case that the scale needs to be changed again if you are moving the particle.

I haven't looked much at the particle code though, so I could be wrong. If you know the variable that is causing the issue, you can add a data breakpoint in visual studio whilst stepping through the code. This will cause the debugger to break when the variable is changed.

I can take a closer look at this later today, just let me know how you get on :)

**Comment by rvanner** [comment]
@plemsoft Maybe you can sort this one?

**Comment by plemsoft** [comment]
@rvanner This has been fixed and will be in the next update :) ,FYI:  @dllllllllllb @ZakJudges search "//PE: Particle scale bug fix." for the solution to this.
